src/ui/edit/pattern-editor.c

```diff
@@ -90,77 +90,82 @@ G_DEFINE_TYPE_WITH_CODE (BtPatternEditor, bt_pattern_editor, GTK_TYPE_WIDGET,
 typedef struct _ParamType
 {
   gint chars, columns;
-  gchar *(*to_string_func) (gchar * dest, gfloat value, gint def);
+  gchar *(*to_string_func) (gchar * dest, gint bufsize, gfloat value, gint def);
   guint column_pos[4];
   gfloat scale;
 } ParamType;
 
 //-- helper methods
 
 static gchar *
-to_string_note (gchar * buf, gfloat value, gint def)
+to_string_note (gchar * buf, gint bufsize, gfloat value, gint def)
 {
   static gchar note_names[] = "C-C#D-D#E-F-F#G-G#A-A#B-????????";
   gint note = ((gint) value) & 255, octave;
 
-  if (note == def || note == 0)
-    return "...";
-  if (note == 255)
-    return "off";
-
-  note--;
-  octave = note >> 4;
-  buf[0] = note_names[2 * (note & 15)];
-  buf[1] = note_names[1 + 2 * (note & 15)];
-  buf[2] = '0' + octave;
-  buf[3] = '\0';
+  if (bufsize >= 4) {
```

Comment:
Please use
https://developer.gnome.org/glib/stable/glib-Warnings-and-Assertions.html#g-return-val-if-fail
at the start to avoid indentation.